File: frontend/lib/site-suggestions.ts
import {
  corridorPolygon,
  lineFromCenter,
  lineLengthM,
  polygonAreaSqm,
  rectanglePolygon,
} from "@/lib/geo";
import type { GeoJSONFeature, GeoJSONGeometry, ProjectType } from "@/lib/types";

export interface SiteSuggestion {
  id: string;
  label: string;
  reason: string;
  score: number;
  geometry: GeoJSONGeometry;
  kind: "boundary" | "alignment";
  building_clashes?: string[];
}

function ringFromPolygon(g: GeoJSONGeometry): [number, number][] | null {
  if (g.type !== "Polygon") return null;
  return (g.coordinates as [number, number][][])[0];
}

function nearestRoadBearing(
  lng: number,
  lat: number,
  roads: GeoJSONFeature[],
): number | null {
  let bestDist = Infinity;
  let bestBearing: number | null = null;
  for (const f of roads) {
    if (f.geometry.type !== "LineString") continue;
    const coords = f.geometry.coordinates as [number, number][];
    for (let i = 1; i < coords.length; i++) {
      const mid: [number, number] = [
        (coords[i - 1][0] + coords[i][0]) / 2,
        (coords[i - 1][1] + coords[i][1]) / 2,
      ];
      const d = Math.hypot(mid[0] - lng, mid[1] - lat);
      if (d < bestDist) {
        bestDist = d;
        const [lng1, lat1] = coords[i - 1];
        const [lng2, lat2] = coords[i];
        const dLng = lng2 - lng1;
        const dLat = lat2 - lat1;
        bestBearing = (Math.atan2(dLng, dLat) * 180) / Math.PI;
      }
    }
  }
  return bestDist < 0.02 ? bestBearing : null;
}

function scoreArea(areaSqm: number, targetMin: number, targetMax: number): number {
  if (areaSqm >= targetMin && areaSqm <= targetMax) return 92;
  if (areaSqm < targetMin) return Math.max(55, 90 - ((targetMin - areaSqm) / targetMin) * 30);
  return Math.max(50, 88 - ((areaSqm - targetMax) / targetMax) * 25);
}

export function generateSiteSuggestions(
  lng: number,
  lat: number,
  projectType: ProjectType,
  roads: GeoJSONFeature[] = [],
): SiteSuggestion[] {
  const roadBearing = nearestRoadBearing(lng, lat, roads);
  const alignBearings = roadBearing != null ? [roadBearing, roadBearing + 90] : [0, 90, 45];

  const suggestions: SiteSuggestion[] = [];

  if (projectType === "building") {
    const presets: { w: number; h: number; label: string; reason: string }[] = [
      { w: 60, h: 45, label: "Compact plot", reason: "Suitable for mid-rise massing (~2,700 m²)" },
      { w: 100, h: 80, label: "Standard campus", reason: "Typical institutional / commercial block" },
      { w: 150, h: 120, label: "Large footprint", reason: "Warehouse or multi-block development" },
    ];
    presets.forEach((p, i) => {
      const geom = rectanglePolygon(lng, lat, p.w, p.h);
      const ring = ringFromPolygon(geom)!;
      const area = polygonAreaSqm(ring.slice(0, -1));
      suggestions.push({
        id: `building-${i}`,
        label: p.label,
        reason: p.reason,
        score: scoreArea(area, 2000, 20000),
        geometry: geom,
        kind: "boundary",
      });
    });
  }

  if (projectType === "flyover" || projectType === "road") {
    const length = projectType === "flyover" ? 650 : 500;
    const width = projectType === "flyover" ? 45 : 35;
    alignBearings.forEach((bearing, i) => {
      const geom = corridorPolygon(lng, lat, length, width, bearing);
      const ring = ringFromPolygon(geom)!;
      const area = polygonAreaSqm(ring.slice(0, -1));
      const aligned = roadBearing != null && Math.abs(bearing - roadBearing) < 15;
      suggestions.push({
        id: `corridor-${i}`,
        label: aligned ? "Road-aligned corridor" : `Corridor ${Math.round(bearing)}°`,
        reason: aligned
          ? "Follows nearest mapped road — good for flyover / grade-separator study"
          : "Alternative alignment axis for comparison",
        score: aligned ? 94 : 78 - i * 3,
        geometry: geom,
        kind: "boundary",
      });
      const line = lineFromCenter(lng, lat, length, bearing);
      suggestions.push({
        id: `align-${i}`,
        label: aligned ? "Centerline (road)" : `Centerline ${Math.round(bearing)}°`,
        reason: "Use as road / flyover alignment for AI design",
        score: aligned ? 90 : 72,
        geometry: line,
        kind: "alignment",
      });
      void area;
    });
  }

  if (projectType === "pipeline") {
    [600, 900].forEach((length, i) => {
      const bearing = alignBearings[i] ?? 0;
      const line = lineFromCenter(lng, lat, length, bearing);
      const len = lineLengthM(line.coordinates as [number, number][]);
      suggestions.push({
        id: `pipeline-${i}`,
        label: `${length} m trench run`,
        reason: roadBearing != null && i === 0 ? "Parallel to nearest utility corridor" : "Cross-country alternative",
        score: i === 0 ? 88 : 75,
        geometry: line,
        kind: "alignment",
      });
      const buffer = corridorPolygon(lng, lat, length, 12, bearing);
      suggestions.push({
        id: `pipeline-buffer-${i}`,
        label: `Trench buffer ${length} m`,
        reason: `${len.toFixed(0)} m centerline with 12 m easement`,
        score: 80 - i * 5,
        geometry: buffer,
        kind: "boundary",
      });
    });
  }

  // Universal: site pad around pin
  suggestions.push({
    id: "pad-standard",
    label: "200 m site pad",
    reason: "General study area around map center — works for any project type",
    score: 70,
    geometry: rectanglePolygon(lng, lat, 200, 200),
    kind: "boundary",
  });

  return suggestions.sort((a, b) => b.score - a.score).slice(0, 8);
}

File: backend/app/api/routes/site_suggestions.py
"""Site suggestion API — server-side heuristics with analysis context."""

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.api.routes.projects import get_owned_project
from app.core.project_catalog import project_type_family
from app.core.disclaimer import DISCLAIMER
from app.core.security import get_current_user_id
from app.db.models import SiteAnalysis
from app.db.session import get_db
from app.services.geospatial import spatial_analysis

router = APIRouter(prefix="/api/projects/{project_id}/site-suggestions", tags=["site-suggestions"])


class SiteSuggestionRequest(BaseModel):
    lng: float | None = None
    lat: float | None = None


def _client_style_suggestions(lng: float, lat: float, project_type: str, roads: list) -> list[dict]:
    """Mirror frontend heuristics for API parity."""
    family = project_type_family(project_type)
    suggestions: list[dict] = []
    if family == "building":
        presets = [
            (60, 45, "Compact plot", "Mid-rise massing (~2,700 m²)"),
            (100, 80, "Standard campus", "Institutional / commercial block"),
            (150, 120, "Large footprint", "Warehouse or multi-block"),
        ]
        for i, (w, h, label, reason) in enumerate(presets):
            ring = _rectangle(lng, lat, w, h)
            area = spatial_analysis.polygon_metrics({"type": "Polygon", "coordinates": [ring]})["area_sqm"]
            suggestions.append(
                {
                    "id": f"api-bld-{i}",
                    "label": label,
                    "reason": reason,
                    "score": 92 if 2000 < area < 15000 else 75,
                    "kind": "boundary",
                    "geometry": {"type": "Polygon", "coordinates": [ring]},
                }
            )
    elif family in ("flyover", "road", "pipeline"):
        lengths = [400, 600, 800] if family != "pipeline" else [300, 500, 700]
        for i, length in enumerate(lengths):
            line = _line_from_center(lng, lat, length, 90 if i % 2 else 0)
            suggestions.append(
                {
                    "id": f"api-align-{i}",
                    "label": f"{length} m corridor",
                    "reason": f"Aligned {'E–W' if i % 2 else 'N–S'} corridor",
                    "score": 88 - i * 3,
                    "kind": "alignment",
                    "geometry": line,
                }
            )
    return suggestions[:6]


def _rectangle(lng: float, lat: float, width_m: float, height_m: float) -> list:
    dlat = height_m / 111320
    dlng = width_m / (111320 * max(0.3, abs(__import__("math").cos(__import__("math").radians(lat)))))
    return [
        [lng - dlng / 2, lat - dlat / 2],
        [lng + dlng / 2, lat - dlat / 2],
        [lng + dlng / 2, lat + dlat / 2],
        [lng - dlng / 2, lat + dlat / 2],
        [lng - dlng / 2, lat - dlat / 2],
    ]


def _line_from_center(lng: float, lat: float, length_m: float, bearing_deg: float) -> dict:
    import math

    br = math.radians(bearing_deg)
    dlat = (length_m / 2) * math.cos(br) / 111320
    dlng = (length_m / 2) * math.sin(br) / (111320 * max(0.3, abs(math.cos(math.radians(lat)))))
    return {
        "type": "LineString",
        "coordinates": [[lng - dlng, lat - dlat], [lng + dlng, lat + dlat]],
    }


def _building_clashes(suggestion_geom: dict, buildings: list) -> list[str]:
    if suggestion_geom.get("type") != "Polygon":
        return []
    clashes = spatial_analysis.find_intersections(suggestion_geom, buildings)
    return [f"Overlaps {c.get('name', c.get('category', 'building'))}" for c in clashes]


@router.post("")
def suggest_sites(
    project_id: int,
    payload: SiteSuggestionRequest | None = None,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
):
    project = get_owned_project(project_id, db, user_id)
    payload = payload or SiteSuggestionRequest()
    lng = payload.lng if payload.lng is not None else (project.center_lng or 77.5946)
    lat = payload.lat if payload.lat is not None else (project.center_lat or 12.9716)

    analysis = (
        db.query(SiteAnalysis)
        .filter(SiteAnalysis.project_id == project_id)
        .order_by(SiteAnalysis.created_at.desc())
        .first()
    )
    roads = []
    buildings = []
    if analysis:
        roads = (analysis.nearby_roads_json or {}).get("features", [])
        buildings = (analysis.existing_buildings_json or {}).get("features", [])

    suggestions = _client_style_suggestions(lng, lat, project.project_type, roads)
    for s in suggestions:
        s["building_clashes"] = _building_clashes(s["geometry"], buildings)

    return {"suggestions": suggestions, "disclaimer": DISCLAIMER}
